### PR TITLE
[FIX #4751] Rails/HasManyOrHasOneDependent shouldn't report offense if :through option was specified

### LINKED ISSUE #4751: Rails/HasManyOrHasOneDependent flagging has_many:through associations
Rails/HasManyOrHasOneDependent seems to be wrongly flagging `has_many through:` associations.

I have the following model
```ruby
class EmailConfig < ApplicationRecord
  has_many :filterings, dependent: :destroy
  has_many :filters, through: :filterings
end
```
--------

## Expected behavior

I expect that running `rubocop` on this file would not report any offense.

## Actual behavior

I get the following report:
```
app/models/email_config.rb:3:3: C: Rails/HasManyOrHasOneDependent: Specify a :dependent option.
  has_many :filters, through: :filterings
  ^^^^^^^^
```

## Steps to reproduce the problem

Run `rubocop` on a fail containing the code above.

## RuboCop version

```
$ rubocop -V
0.50.0 (using Parser 2.4.0.0, running on ruby 2.4.2 x86_64-darwin16)
```

### PR DESCRIPTION
**Pull request by smakagon**
Without this fix, `Rails/HasManyOrHasOneDependent` would register offense for this code:

```ruby
class EmailConfig < ApplicationRecord
  has_many :filterings, dependent: :destroy
  has_many :filters, through: :filterings
end
```
With this fix, it doesn't register offense if `:through` option was passed.

-----------------

Before submitting the PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it).
* [x] Squashed related commits together.
* [x] Added tests.
* [x] Added an entry to the [Changelog](../blob/master/CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../blob/master/CONTRIBUTING.md#changelog-entry-format).
* [x] All tests(`rake spec`) are passing.
* [x] The new code doesn't generate RuboCop offenses that are checked by `rake internal_investigation`.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [x] Updated cop documentation with `rake generate_cops_documentation` (required only when you've added a new cop or changed the configuration/documentation of an existing cop).

[1]: http://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: lib/rubocop/cop/rails/has_many_or_has_one_dependent.rb [hunk 1]
```diff
@@ -33,11 +33,17 @@ class HasManyOrHasOneDependent < Cop
           (pair (sym :dependent) !(:nil))
         PATTERN
 
+        def_node_matcher :has_through?, <<-PATTERN
+          (pair (sym :through) !(:nil))
```
Comment (Drenmi) [inline lib/rubocop/cop/rails/has_many_or_has_one_dependent.rb]: It's possible to do `has_many ..., through: :nil`? If this is an edge case, let's add a test for it. 🙂

Comment (smakagon) [reply lib/rubocop/cop/rails/has_many_or_has_one_dependent.rb]: Extended specs and mentioned `:through` option in the documentation for the cop.

File: spec/rubocop/cop/rails/has_many_or_has_one_dependent_spec.rb [hunk 2]
```diff
@@ -49,5 +59,15 @@ class Person
...
       expect_no_offenses('has_many :foo, dependent: :bar')
     end
+
+    context 'with :through option' do
+      it 'does not register an offense for non-nil value' do
+        expect_no_offenses('has_many :foo, through: :bars')
+      end
+
+      it 'registers an offense for nil value' do
+        expect_no_offenses('has_many :foo, through: :nil')
```
Comment (Drenmi) [inline spec/rubocop/cop/rails/has_many_or_has_one_dependent_spec.rb]: Where is this behaviour documented though? I've never seen `through: :nil`, can't figure out what it would do, nor find any documentation for it. 🤔

File: lib/rubocop/cop/rails/has_many_or_has_one_dependent.rb (hunk 1 above)
Comment (smakagon) [reply lib/rubocop/cop/rails/has_many_or_has_one_dependent.rb]: @Drenmi  I've checked, `through: nil` is a possible case and Rails will not complain about it. But it will throw an error once you try to use that relation in an application. So I think it makes sense to have this check.

**Comment by smakagon** [comment]
@Drenmi seems like this PR is ready to go. I've rebased it against a master and made all updates that you've suggested.

**Comment by oboxodo** [comment]
Woohoo! Thanks for this patch! We actually discussed this caveat when I originally developed the cop (#4541) and it was pending a solution when I went on vacation. Thanks for finishing this!
